app.py

```diff
@@ -879,7 +912,12 @@ def replacesubtree():
 	sent = SENTENCES[QUEUE[sentno - 1][0]]
 	orig_senttok, _ = worker.postokenize(sent)
 	username = session['username']
-	treestr = request.args.get('tree', '')
+	treestr = request.args.get('tree')
+	if app.config['CGELVALIDATE'] is None:
+		treestr = request.args.get('tree')
+	else: 
+		treestr = "(ROOT " + cgel.parse(request.args.get('tree'))[0].ptb() + ")"
+		treestr = writediscbrackettree(DrawTree(treestr).nodes[0],orig_senttok)
```

Comment:
A lot of the new blocks are some variant of this. Is there a reason to repeat in each function or is this a good opportunity for refactoring? (At a glance it's hard to see the various ways that these functions differ in terms of state. I think that's due to the way the code was originally written, so maybe we should do a bigger refactoring later.)

Reply:
It's the latter -- this would be a great opportunity for eventual refactoring. (The redundancies in the new blocks reflect redundancies in the original code)